qa/tasks/quiescer.py

```diff
@@ -0,0 +1,236 @@
+                self.stopping.wait(5)
+
+        while leader is not None and not self.stopping.is_set():
+            # We use the one_shot here to cover for cases when the mds crashes
+            # without this parameter the client may get stuck awaiting response from a dead MDS
+            command = ['tell', f"mds.{leader}", 'quiesce', 'db']
+            command.extend(*args)
+            self.logger.info("Running ceph command: '%s'" % " ".join(command))
+            result = self.fs.run_ceph_cmd(args=command, stdout=StringIO())
+            rc, stdout = result.exitstatus, result.stdout.getvalue()
+            if rc == -errno.ENOTTY:
+                try:
+                    resp = json.loads(stdout)
+                    leader = int(resp['leader'])
+                    self.logger.info("Retrying a quiesce db command with leader %d" % leader)
+                except Exception as e:
+                    self.logger.error("Couldn't parse ENOTTY response from an mds with error: %s\n%s" % (str(e), stdout))
+                    self.stopping.wait(5)
+            else:
+                break
+
+        return (rc, stdout)
+
+    
+    def do_quiesce(self, duration):
+
+        # quiesce the root
+        rc, stdout = self.tell_quiesce_leader(
+            "/", # quiesce at the root
+            "--timeout", str(self.quiesce_timeout), 
+            "--expiration", str(duration + 60), # give us a minute to run the release command
+            "--await" # block until quiesced (or timedout)
+        )
+
+        if self.stopping.is_set():
+            return
+        
+        if rc != 0:
+            rcinfo = f"{-rc} ({errno.errorcode.get(-rc, 'Unknown')})"
+            self.logger.error(f"Coulding quiesce root with rc: {rcinfo}, stdout:\n{stdout}")
+            raise RuntimeError(f"Error quiescing root: {rcinfo}")
+        
+        try:
+            response = json.loads(stdout)
+            set_id = response["sets"].keys()[0]
+        except Exception as e:
+            self.logger.error(f"Coulding parse response with error {e}; stdout:\n{stdout}")
+            raise RuntimeError(f"Error parsing quiesce response: {e}")
+
+        self.logger.info(f"Successfully quiesced, set_id: {set_id}, quiesce duration: {duration}")
+        self.stopping.wait(duration)
+
+        # release the root
+        rc, stdout = self.tell_quiesce_leader(
+            "--set-id", set_id,
+            "--release",
+            "--await"
+        )
+        
+        if rc != 0:
+            rcinfo = f"{-rc} ({errno.errorcode.get(-rc, 'Unknown')})"
+            self.logger.error(f"Coulding release root with rc: {rcinfo}, stdout:\n{stdout}")
+            raise RuntimeError(f"Error releasing root: {rcinfo}")
+
+
+    def _run(self):
+        try:
+            self.stopping.wait(self.initial_delay)
+
+            while not self.stopping.is_set():
+                duration = self.next_quiesce_duration()
+                self.do_quiesce(duration)
```

Comment:
Keep in mind that this could be run concurrently with the MDSThrasher. While you don't need ot account for that righ tnow but be aware that a rank or ranks may go down and cause a quiesce to be cancelled.

Reply:
Handle it gracefully. We would probably adjust the configs of both so that they do not overlap constantly.

Reply:
This could be tricky. I can ignore errors from release when a set is CANCELED, would that suffice?

Reply:
I don't think these two trashers are compatible. How should we tell an MDS crash caused by the quiesce and leading to CANCELED quiesce set from an MDS crasher induced rank failure?

Reply:
> I don't think these two trashers are compatible. How should we tell an MDS crash caused by the quiesce and leading to CANCELED quiesce set from an MDS crasher induced rank failure?

MDSThrasher by default lets `daemon-helper` send `SIGKILL` to the MDS. You won't be able to see that return code because the processes is waited for after sending the signal.

In general, your thrasher shouldn't care about crashes at all. The `DaemonWatchdog` will stop the tests if one of the MDS is dead for longer than 300 seconds (IIRC). The log/core examination at the end of the test will also detect an inappropriate crash.

Perhaps some introspective information can be added to explain why the set was cancelled. If an MDS failover occurred, that's reasonable (for your thrasher's perspective). You can compare `FSStatus` to detect this using `FSStatus.hadfailover`.

Reply:
> In general, your thrasher shouldn't care about crashes at all.

Oh ok, here is the root of my confusion. The way I structured the thrasher was to actively report failure if I couldn't release the pause successfully. Are you suggesting that I don't do it, and instead rely on the other running workloads to fail as needed?

Reply:
> > In general, your thrasher shouldn't care about crashes at all.
> 
> Oh ok, here is the root of my confusion. The way I structured the thrasher was to actively report failure if I couldn't release the pause successfully.

I think raising an error from the thrasher is reasonable if (a) timeout was exceeded; (b) no MDS failover occurred.

You should also be robust to an MDS tell command timing out due to failover. Unless no failover occurred, then raise an error.

> Are you suggesting that I don't do it, and instead rely on the other running workloads to fail as needed?

If by "workloads" you mean other "failure detection mechanisms", yes, but specifically for "MDS failover occurred".

Reply:
Eventually, I am not considering a CANCELed set as an error, neither during quiesce nor during release. The condition is warned about.